stablehlo/dialect/TypeInference.cpp

```diff
@@ -1484,8 +1482,57 @@ LogicalResult inferCholeskyOp(
         location, "minor dimensions of 'a' must have equal size, got shape ",
         aShape, ".");
   }
-  inferredReturnShapes.emplace_back(aRankedType.getShape(),
-                                    aRankedType.getElementType());
+
+  // bound infer rules for the last two dimensions of A:
+  //        Dim R-2 | Dim R-1 | inferred R-2 | inferred R-1
+  // case0: 3       | 3       | 3            | 3
+  // case1: ?       | 3       | 3            | 3
+  // case2: ?, A>=3 | 3       | 3            | 3
+  // case3: 3       | ?       | 3            | 3
+  // case4: 3       | ?, A>=3 | 3            | 3
+  auto resultBounds = encodingToBounds(aRankedType.getEncoding()).vec();
+  if (isStaticDimSize(lastDim) || isStaticDimSize(penultimateDim)) {
+    auto resultShape = aRankedType.getShape().vec();
+    auto staticSize = isStaticDimSize(lastDim) ? lastDim : penultimateDim;
+    resultShape[resultShape.size() - 1] = staticSize;
+    resultShape[resultShape.size() - 2] = staticSize;
+    if (resultBounds.empty()) {
+      inferredReturnTypes.push_back(
+          RankedTensorType::get(resultShape, aRankedType.getElementType()));
+    } else {
+      resultBounds[resultBounds.size() - 1] = ShapedType::kDynamic;
+      resultBounds[resultBounds.size() - 2] = ShapedType::kDynamic;
+      inferredReturnTypes.push_back(RankedTensorType::get(
+          resultShape, aRankedType.getElementType(),
+          boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
+    }
+    return success();
+  }
+  // case5: ?       | ?       | ?            | ?
+  if (resultBounds.empty()) {
+    inferredReturnTypes.push_back(RankedTensorType::get(
+        aRankedType.getShape(), aRankedType.getElementType(),
+        boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
+    return success();
+  }
+  // case6: ?, A    | ?       | ?, A         | ?, A
+  // case7: ?       | ?, A    | ?, A         | ?, A
+  // case8: ?, A    | ?, B    | ?, min(A,B)  | ?, min(A,B)
+  bool isLastDimStaticBound =
+      !isDynamicDimSize(resultBounds[resultBounds.size() - 1]);
+  bool isPenultimateDimStaticBound =
+      !isDynamicDimSize(resultBounds[resultBounds.size() - 2]);
+  int64_t staticBound = isLastDimStaticBound
+                            ? resultBounds[resultBounds.size() - 1]
+                            : resultBounds[resultBounds.size() - 2];
+  if (isLastDimStaticBound && isPenultimateDimStaticBound)
+    staticBound = std::min(resultBounds[resultBounds.size() - 1],
+                           resultBounds[resultBounds.size() - 2]);
+  resultBounds[resultBounds.size() - 1] = staticBound;
+  resultBounds[resultBounds.size() - 2] = staticBound;
+  inferredReturnTypes.push_back(RankedTensorType::get(
+      aRankedType.getShape(), aRankedType.getElementType(),
+      boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
```

Comment:
This is a lot of code - can we make this simpler? E.g. can we use `inferMostSpecificSomething` here instead?

Reply:
We can't :( `inferMostSpecificXXX` functions are inferring using variadic types and using information from multiple types to find a single most-specific type, but since this op is dealing with tightening a single type, it doesn't fit our use case. I've actually made some significant changes since yesterday so it requires rereview. This op was confusing at first, but after figuring out that some rules are not bound related, things made much more sense.